src/kernel.ts

```diff
@@ -1,1069 +1,784 @@
+    if (this._ws === null) {
+      return false;
+    }
+    if (this._ws.readyState !== WebSocket.OPEN) {
+      return false;
+    }
+    return true;
+  }
+
+  /**
+   * Check whether the connection to the kernel has been completely
+   * severed. This function only returns true if all channel objects
+   * are null.
+   */
+  isFullyDisconnected(): boolean {
+    return (this._ws === null);
+  }
+
+  /**
+   * Send a message on the Kernel's shell channel.
+   */
+  sendShellMessage(msg_type: string, content: any, metadata = {}, buffers: string[] = []): IKernelFuture {
+    if (!this.isConnected()) {
+      throw new Error("kernel is not connected");
+    }
+    var msg = this._getMsg(msg_type, content, metadata, buffers);
+    msg.channel = 'shell';
+
+    this._ws.send(serialize.serialize(msg));
+
+    var future = new KernelFutureHandler(() => {
+      this._handlerMap.delete(msg.header.msgId);
+    });
+
+    this._handlerMap.set(msg.header.msgId, future);
+
+    return future;
+  }
+
+  /**
+   * Get kernel info.
+   *
+   * Returns a KernelFuture that will resolve to a `kernel_info_reply` message documented
+   * [here](http://ipython.org/ipython-doc/dev/development/messaging.html#kernel-info)
+   */
+  kernelInfo(): IKernelFuture {
+    return this.sendShellMessage("kernel_info_request", {});
+  }
+
+  /**
+   * Get info on an object.
+   *
+   * Returns a KernelFuture that will resolve to a `inspect_reply` message documented
+   * [here](http://ipython.org/ipython-doc/dev/development/messaging.html#object-information)
+   */
+  inspect(code: string, cursor_pos: number): IKernelFuture {
+    var content = {
+      code: code,
+      cursor_pos: cursor_pos,
+      detail_level: 0
+    };
+    return this.sendShellMessage("inspect_request", content);
+  }
+
+  /**
+   * Execute given code into kernel, returning a KernelFuture.
+   *
+   * @example
+   *
+   * The options object should contain the options for the execute
+   * call. Its default values are:
+   *
+   *      options = {
+   *        silent : true,
+   *        user_expressions : {},
+   *        allow_stdin : false
+   *      }
+   *
+   */
+  execute(code: string, options?: { silent?: boolean; user_expressions?: any; allow_stdin?: boolean; }): IKernelFuture {
+    var content = {
+      code: code,
+      silent: true,
+      store_history: false,
+      user_expressions: {},
+      allow_stdin: false
+    };
+    utils.extend(content, options);
+    return this.sendShellMessage("execute_request", content);
+  }
+
+  /**
+   * Request a code completion from the kernel.
+   *
+   * Returns a KernelFuture with will resolve to a `complete_reply` documented
+   * [here](http://ipython.org/ipython-doc/dev/development/messaging.html#complete)
+   */
+  complete(code: string, cursor_pos: number): IKernelFuture {
+    var content = {
+      code: code,
+      cursor_pos: cursor_pos
+    };
+    return this.sendShellMessage("complete_request", content);
```

Comment:
At some point, typescript may get string literal types, which would allow creating a string literal union, which will be awesome and solve this problem. https://github.com/Microsoft/TypeScript/issues/1003